prov/gni/src/gnix_ep.c

```diff
@@ -182,25 +479,108 @@ static ssize_t gnix_ep_recvmsg(struct fid_ep *ep, const struct fi_msg *msg,
 
 }
 
-static const send_func_t const send_method[] = {
-					[FI_EP_MSG] =  gnix_ep_send_msg,
-					[FI_EP_RDM] =  gnix_ep_send_rdm,
-					};
-
 static ssize_t gnix_ep_send(struct fid_ep *ep, const void *buf, size_t len,
 			    void *desc, fi_addr_t dest_addr, void *context)
 {
-	struct gnix_fid_ep *ep_priv;
+	int ret = FI_SUCCESS;
+	struct gnix_vc *vc = NULL;
+	struct gnix_nic *nic = NULL;
+	struct gnix_fid_ep *ep_priv = NULL;
+	struct gnix_fab_req *req;
+	struct gnix_address *addr_ptr;
+	struct gnix_tx_descriptor *tdesc;
+	gni_return_t status;
 
 	ep_priv = container_of(ep, struct gnix_fid_ep, ep_fid);
-	assert((ep_priv->type == FI_EP_RDM) || (ep_priv->type == FI_EP_MSG));
 
-	return send_method[ep_priv->type](ep,
-					  buf,
-					  len,
-					  desc,
-					  dest_addr,
-					  context);
+	nic = ep_priv->nic;
+	if (nic == NULL)
+		return -FI_EINVAL;
+
+	if (ep_priv->type == FI_EP_RDM) {
+		ret = _gnix_ep_get_vc(ep_priv, dest_addr, &vc);
+		if (unlikely(ret != FI_SUCCESS)) {
+			GNIX_WARN(FI_LOG_EP_DATA,
+				  "_gnix_ep_get_vc returned %d\n",
+				   ret);
+			goto err;
+		}
+	} else
+		vc = ep_priv->vc; /* FI_EP_MSG easy */
+
+	if (likely(vc->conn_state == GNIX_VC_CONNECTED) &&
+			slist_empty(&vc->send_queue)) {
+		if (len < (16384 - sizeof(struct gnix_smsg_hdr))) {
+			/*
+			 * pure smsg path
+			 */
+			ret = _gnix_nic_tx_alloc(ep_priv->nic,
+						 &tdesc);
```

Comment:
So you don't use a gnix_fab_req in the fast path and just keep context and other data in the TXD.  In my RMA work I always allocated a gnix_fab_req to encapsulate the TXD structure and hold context and state stuff.  We may want to change one path or the other to match